### PR TITLE
Add -Delimiter to ConvertFrom-StringData

### PR DESCRIPTION
**Pull request by steviecoaster**
# PR Summary

This PR adds the ability to specify a delimiter to the ConvertFrom-StringData cmdlet

## PR Context

ConvertFrom-StringData is powerful, but not very discoverable. This addition eases usability of the cmdlet to work with many string formats easily.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/4864
- **Testing - New and feature**
    - [] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-StringData.cs [hunk 1]
```diff
@@ -30,19 +30,34 @@ public string StringData
...
                 return _stringData;
             }
+
             set
             {
                 _stringData = value;
             }
         }
 
+        private string _Delimiter;
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-StringData.cs]: Please remove private field and use auto-implemented property
```c#
 public char Delimiter { get; set; } = '=';

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-StringData.cs [hunk 2]
```diff
@@ -30,19 +30,34 @@ public string StringData
...
+
+        /// <summary>
+        /// The delimiter to index from to create the object
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-StringData.cs]: Should starts with "Gets or sets" and has final dot.
```suggestion
        /// Gets or sets the delimiter.
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-StringData.cs [hunk 3]
```diff
@@ -30,19 +30,34 @@ public string StringData
...
+        /// </summary>
+        /// <value></value>
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-StringData.cs]: Please remove.
```suggestion
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-StringData.cs [hunk 4]
```diff
@@ -54,10 +69,10 @@ protected override void ProcessRecord()
             {
                 string s = line.Trim();
 
-                if (String.IsNullOrEmpty(s) || s[0] == '#')
+                if (string.IsNullOrEmpty(s) || s[0] == '#')
                     continue;
 
-                int index = s.IndexOf('=');
+                int index = s.IndexOf(Delimiter);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-StringData.cs]: It was a char. Now it is string. I believe it should be char like csv cmdlets have.

**Review by iSazonov — COMMENTED** [review]
Please add new tests and update existing tests.

You should sigh CLA before we can merge the PR.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-StringData.cs (hunk 2 above)
Comment (steviecoaster) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-StringData.cs]: Addressed in most recent commit

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-StringData.cs (hunk 1 above)
Comment (steviecoaster) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-StringData.cs]: Addressed in most recent commit

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-StringData.cs (hunk 3 above)
Comment (steviecoaster) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-StringData.cs]: Addressed in most recent commit

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-StringData.cs (hunk 4 above)
Comment (steviecoaster) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFrom-StringData.cs]: Addressed in most recent commit

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1 [hunk 5]
```diff
@@ -65,3 +65,50 @@ bazz = 2
 	$(ConvertFrom-StringData -StringData $sampleData).Values | Should -BeIn @("0","1","2")
     }
 }
+
+Describe "Delimiter parameter tests" -Tags "CI" {
+
+    $sampleData = @"
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1]: It is used only in one test. Please move to the test.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1 [hunk 6]
```diff
@@ -65,3 +65,50 @@ bazz = 2
...
+
+a:b
+
+"@
+
+
+    It "Should return the data on the left side in the key" {
+        $actualValue = ConvertFrom-StringData -StringData 'a:b' -Delimiter ':'
+
+        $actualValue.Keys | Should -BeExactly "a"
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1]: Test below with TestCases does the check so we can remove the test.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1 [hunk 7]
```diff
@@ -65,3 +65,50 @@ bazz = 2
...
+    It "Should not throw on given delimiter" {
+    {$sampleData | ConvertFrom-StringData -Delimiter ':' }| Should -Not -Throw
+    }
+
+
+    $TestCases = @(
+        @{ Delimiter = ':'; StringData = 'value:10'; ExpectedResult = @{ Values = 10 } }
+        @{ Delimiter = '-'; StringData = 'a-b' ; ExpectedResult = @{ Values = 'b' } }
+        @{ Delimiter = ','; StringData = 'c,d' ; ExpectedResult = @{ Values = 'd' } }
+    )
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1]: Please put in BeforeAll block.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1 [hunk 8]
```diff
@@ -65,3 +65,50 @@ bazz = 2
...
+        $actualValue = ConvertFrom-StringData -StringData 'a:b' -Delimiter ':'
+
+        $actualValue.Keys | Should -BeExactly "a"
+    }
+
+    It "Should return the data on the right side in the value" {
+        $actualValue = ConvertFrom-StringData -StringData 'a=b'
+
+        $actualValue.Values | Should -BeExactly "b"
+    }
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1]: Really the test checks default delimiter. Please update title ("Default delimiter '=' works") and add check for key ` $actualValue.Keys | Should -BeExactly "a"`.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1 [hunk 9]
```diff
@@ -65,3 +65,50 @@ bazz = 2
...
+        @{ Delimiter = '-'; StringData = 'a-b' ; ExpectedResult = @{ Values = 'b' } }
+        @{ Delimiter = ','; StringData = 'c,d' ; ExpectedResult = @{ Values = 'd' } }
+    )
+
+    It 'is able to parse <StringData> with delimiter "<Delimiter>"' -TestCases $TestCases {
+        param($Delimiter, $StringData, $ExpectedResult)
+
+        $Result = ConvertFrom-StringData -StringData $StringData -Delimiter $Delimiter
+        foreach ($Name in $ExpectedResult.Keys) {
+            $Result.$Name | Should -Be $ExpectedResult.$Name
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1]: Please add value check.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1 [hunk 10]
```diff
@@ -65,3 +65,50 @@ bazz = 2
...
+        }
+
+    }
+
+
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1]: Please remove extra lines.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1 [hunk 11]
```diff
@@ -65,3 +65,50 @@ bazz = 2
...
+}
+
+
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1]: The same.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1 (hunk 9 above)
Comment (steviecoaster) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1]: This is actually checking both name _and_ value, at the same time. It just reads funny with that `$Name` Variable

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1 [hunk 12]
```diff
@@ -65,3 +65,37 @@ bazz = 2
...
+        $actualValue.Values | Should -BeExactly "b"
+        $actualValue.Keys | Should -BeExactly "a"
+    }
+
+    It "Should not throw on given delimiter" {
+        $sampleData = @"
+a:b
+"@
+    { $sampleData | ConvertFrom-StringData -Delimiter ':' }| Should -Not -Throw
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1]:
```suggestion
    { $sampleData | ConvertFrom-StringData -Delimiter ':' } | Should -Not -Throw
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1 (hunk 9 above)
Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1]: It looks a tricky. Could you please make the test more clear and explicitly check expected keys and value?

Comment (vexx32) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1]: Maybe something like this?
```powershell
$Name | Should -BeIn $Result.Keys
$Result.$Name | Should -Be $ExpectedResult.$Name
```

And then you could also rename `$Name` to `$Key` or something so it's slightly clearer.

Comment (steviecoaster) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1]: Done!

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1 (hunk 10 above)
Comment (steviecoaster) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1]: Done!

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1 (hunk 11 above)
Comment (steviecoaster) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1]: Done!

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1 (hunk 12 above)
Comment (steviecoaster) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1]: Done!

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1 (hunk 8 above)
Comment (steviecoaster) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1]: Done!

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1 (hunk 7 above)
Comment (steviecoaster) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1]: Done!

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1 (hunk 6 above)
Comment (steviecoaster) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1]: Done!

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1 (hunk 5 above)
Comment (steviecoaster) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1]: Done!

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1 [hunk 13]
```diff
@@ -65,3 +65,39 @@ bazz = 2
...
+
+        $actualValue.Values | Should -BeExactly "b"
+        $actualValue.Keys | Should -BeExactly "a"
+    }
+
+    It "Should not throw on given delimiter" {
+        $sampleData = @"
+a:b
+"@
+    { $sampleData | ConvertFrom-StringData -Delimiter ':' } | Should -Not -Throw
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1]: Please fix indentation.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1 [hunk 14]
```diff
@@ -65,3 +65,39 @@ bazz = 2
...
+        param($Delimiter, $StringData, $ExpectedResult)
+
+        $Result = ConvertFrom-StringData -StringData $StringData -Delimiter $Delimiter
+
+        foreach ($Key in $ExpectedResult.Keys) {
+            $Key | Should -BeIn $ExpectedResult.Keys
+            $Result.$Key | Should -Be $ExpectedResult.$Key
+        }
+    }
+
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-StringData.Tests.ps1]: Please remove extra line.

**Review by iSazonov — APPROVED** [review]
LGTM with two style comments.

**Comment by anmenaga** [comment]
CI passed on rebuild.

**Review by anmenaga — APPROVED** [review]

**Comment by vexx32** [comment]
Nice work @steviecoaster! 💖

**Comment by iSazonov** [comment]
@steviecoaster Thanks for your contribution!

**Comment by chuanjiao10** [comment]
-Delimiter If it can be string, it is best ！！！

**Pull request merged**